kornia/grad_estimator/ste.py

```diff
@@ -0,0 +1,105 @@
+from typing import Callable, Optional, Tuple
+
+import torch.nn as nn
+from torch import Tensor
+from torch.autograd import Function
+
+__all__ = [
+    "STEFunction",
+    "StraightThroughEstimator"
+]
+
+
+class STEFunction(Function):
+    """Straight-Through Estimation (STE) function.
+
+    STE bridges the gradients between the input tensor and output tensor as if the function
+    was an identity function. Meanwhile, advanced gradient functions are also supported. e.g.
+    the output gradients can be mapped into [-1, 1] with ``F.hardtanh`` function.
+
+    Args:
+        grad_fn: function to restrain the gradient recieved. If None, no mapping will performed.
+
+    Example:
+        Let the gradients of ``torch.sign`` estimated from STE.
+        >>> input = torch.randn(4, requires_grad = True)
+        >>> output = torch.sign(input)
+        >>> loss = output.mean()
+        >>> loss.backward()
+        >>> input.grad
+        tensor([0., 0., 0., 0.])
+
+        >>> with torch.no_grad():
+        ...     output = torch.sign(input)
+        >>> out_est = STEFunction.apply(input, output)
+        >>> loss = out_est.mean()
+        >>> loss.backward()
+        >>> input.grad
+        tensor([0.2500, 0.2500, 0.2500, 0.2500])
+    """
+
+    @staticmethod
+    def forward(  # type:ignore
+        ctx, input: Tensor, output: Tensor,
+        grad_fn: Optional[Callable] = None
+    ) -> Tensor:
+        ctx.in_shape = input.shape
+        ctx.out_shape = output.shape
+        ctx.grad_fn = grad_fn
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output: Tensor) -> Tuple[Tensor, Tensor, None]:  # type:ignore
```

Comment:
will this work with onnx/jit ? usually ignores doesn't support well tracing

Reply:
torch.Function does not support well with ONNX and JIT. Probably need to have a suite of C++ code for that.

https://pytorch.org/docs/1.10.0/onnx.html#torch-autograd-functions

https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html